File: native/engine-chromium-desktop/src/handlers/console_handler.cpp
#include "console_handler.h"

namespace mollotov {

ConsoleHandler::ConsoleHandler(DesktopHandlerRuntime runtime) : runtime_(std::move(runtime)) {}

void ConsoleHandler::Register(DesktopRouter& router) const {
  router.Register("get-console-messages",
                  [this](const nlohmann::json& params) { return GetConsoleMessages(params); });
  router.Register("get-js-errors", [this](const nlohmann::json&) { return GetJsErrors(); });
  router.Register("clear-console", [this](const nlohmann::json&) { return ClearConsole(); });
}

nlohmann::json ConsoleHandler::GetConsoleMessages(const nlohmann::json& params) const {
  if (runtime_.console_store == nullptr) {
    return SuccessResponse({{"messages", nlohmann::json::array()}, {"count", 0}, {"hasMore", false}});
  }
  const auto level_it = params.find("level");
  const std::optional<std::string> level =
      level_it != params.end() && level_it->is_string()
          ? std::optional<std::string>(level_it->get<std::string>())
          : std::nullopt;
  const int limit = std::max(1, IntOrDefault(params, "limit", 100));
  nlohmann::json entries = ParseJsonText(runtime_.console_store->ToJson(level));
  if (entries.size() > static_cast<std::size_t>(limit)) {
    entries.erase(entries.begin(), entries.begin() + static_cast<long>(entries.size() - limit));
  }
  return SuccessResponse({
      {"messages", entries},
      {"count", entries.size()},
      {"hasMore", runtime_.console_store->Count() > static_cast<int>(entries.size())},
  });
}

nlohmann::json ConsoleHandler::GetJsErrors() const {
  if (runtime_.console_store == nullptr) {
    return SuccessResponse({{"errors", nlohmann::json::array()}, {"count", 0}});
  }
  nlohmann::json errors = ParseJsonText(runtime_.console_store->GetErrorsOnly());
  for (auto& entry : errors) {
    entry["type"] = "console-error";
  }
  return SuccessResponse({{"errors", errors}, {"count", errors.size()}});
}

nlohmann::json ConsoleHandler::ClearConsole() const {
  if (runtime_.console_store == nullptr) {
    return SuccessResponse({{"cleared", 0}});
  }
  const int count = runtime_.console_store->Count();
  runtime_.console_store->Clear();
  return SuccessResponse({{"cleared", count}});
}

}  // namespace mollotov

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/ConsoleHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.network.Router
import com.kelpie.browser.network.successResponse

class ConsoleHandler(
    private val context: HandlerContext,
) {
    fun register(router: Router) {
        router.register("get-console-messages") { body -> getConsoleMessages(body) }
        router.register("get-js-errors") { getJSErrors() }
        router.register("clear-console") { clearConsole() }
    }

    private fun getConsoleMessages(body: Map<String, Any?>): Map<String, Any?> {
        val level = body["level"] as? String
        val limit = body["limit"] as? Int ?: 100

        var messages = context.snapshotConsoleMessages()
        if (level != null) {
            messages = messages.filter { it["level"] == level }
        }
        val limited = messages.takeLast(limit)
        return successResponse(
            mapOf(
                "messages" to limited,
                "count" to limited.size,
                "hasMore" to (messages.size > limit),
            ),
        )
    }

    private fun getJSErrors(): Map<String, Any?> {
        val errors =
            context.snapshotConsoleMessages().filter { it["level"] == "error" }
        return successResponse(
            mapOf(
                "errors" to errors.map { it + ("type" to "console-error") },
                "count" to errors.size,
            ),
        )
    }

    private fun clearConsole(): Map<String, Any?> = successResponse(mapOf("cleared" to context.clearConsoleMessages()))

    companion object {
        const val BRIDGE_SCRIPT =
            """
            (function() {
                var _origConsole = {};
                ['log','warn','error','info','debug'].forEach(function(level) {
                    _origConsole[level] = console[level];
                    console[level] = function() {
                        var args = Array.from(arguments).map(function(a) {
                            try { return typeof a === 'object' ? JSON.stringify(a) : String(a); }
                            catch(e) { return String(a); }
                        });
                        var msg = {
                            level: level,
                            text: args.join(' '),
                            timestamp: new Date().toISOString(),
                            source: '',
                            line: 0,
                            column: 0,
                            stackTrace: null
                        };
                        try {
                            var stack = new Error().stack || '';
                            var lines = stack.split('\n');
                            if (lines.length > 2) {
                                var match = lines[2].match(/(https?:\/\/[^:]+):?(\d+)?:?(\d+)?/);
                                if (match) {
                                    msg.source = match[1] || '';
                                    msg.line = parseInt(match[2]) || 0;
                                    msg.column = parseInt(match[3]) || 0;
                                }
                                msg.stackTrace = lines.slice(2).join('\n');
                            }
                        } catch(e) {}
                        if (window.KelpieBridge) window.KelpieBridge.onConsoleMessage(JSON.stringify(msg));
                        _origConsole[level].apply(console, arguments);
                    };
                });
                window.addEventListener('error', function(e) {
                    if (window.KelpieBridge) window.KelpieBridge.onConsoleMessage(JSON.stringify({
                        level: 'error',
                        text: e.message || String(e),
                        source: e.filename || '',
                        line: e.lineno || 0,
                        column: e.colno || 0,
                        timestamp: new Date().toISOString(),
                        stackTrace: e.error ? e.error.stack : null
                    }));
                });
                window.addEventListener('unhandledrejection', function(e) {
                    if (window.KelpieBridge) window.KelpieBridge.onConsoleMessage(JSON.stringify({
                        level: 'error',
                        text: 'Unhandled Promise rejection: ' + (e.reason ? (e.reason.message || String(e.reason)) : 'unknown'),
                        source: '',
                        line: 0,
                        column: 0,
                        timestamp: new Date().toISOString(),
                        stackTrace: e.reason ? e.reason.stack : null
                    }));
                });
            })();
            """
    }
}
